File: L1_merge_sort.py
#Imports
from array import array
import random as r

# Constants
N = 10

# Globals
x = array('i', [0]*N)

# Subroutines
def merge(A, p, q, r):

    if type(A) is list:
        left = A[p:q+1]
        right = A[q+1: r+1]
    else:
        left =      A[p:q+1]
        right =      A[q+1: r+1]

    i = 0
    j = 0
    k = p

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            A[k] = left[i]
            i += 1
        else:
            A[k] = right[j]
            j += 1

        k += 1

    if i < len(left):
        A[k : r+1] = left[i:]
    if j < len(right):
        A[k : r+1] = right[j:]
            
def merge_sort(A, p = 0, r = None):

    if r is None:
        r = len(A) - 1
    if p >= r:
        return
    q = (p+r)//2
    merge_sort(A, p, q)
    merge_sort(A, q+1, r)
    merge(A, p, q, r)

# Printing
def print_list(list, n):
    for i in range(0, n):
        print("%3d" % (list[i]), end="")
    print()

# Main Routine
def main():

    # Initial Header
    print("Lab1 - Merge Sort")
    print()

    # For list to initialize array elements
    for i in range(0, N):
        x[i] = r.randint(0, 99)

    # Print initialized elements and explanatory header
    print("List before sorting:")
    print_list(x, N)
    print()

    # Invoke merge sort algorithm
    merge_sort(x)

    # Print sorted list and explanatory header
    print("List after sorting:")
    print_list(x, N)
    
# Start-up Scripts
if __name__ == "__main__":
    main()

    

File: L1_merge_sort.c
// Imports
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Constants
#define N 250000

// Globals
int x[N];

void merge_sort_O0(int *A, int p, int r);
void merge_sort_O3(int *A, int p, int r);
// Subroutines


// Printing Routine
void print_list(int *list, int n)
{
  for(int i = 0; i<n; i++)
    {
      printf(" %3d", list[i]);
    };
  printf("\n");
}


// Insertion Sort Routine 
void insertion_sort(int *list, int n)
{
  int key;
  int j;
  
  for(int i = 1; i < N; i++)
    {
      key = list[i];
      j = i - 1;

      while((j>=0) && (list[j]>key))
	{
	  list[j +1] = list[j];
	  j -= 1;
	};
      list[j+1] = key;
    };
}

// Main routine
int main(void)
{
  srand(time(NULL));
  printf("Lab1 - Insertion Sort\n\n");

  for(int i = 0; i < N; i++)
    {
      x[i] = (rand()>>24);
    }

//   printf("List before sorting\n");
//   print_list(x, N);

//   printf("\nList after sorting\n");
//   insertion_sort(x, N);
    merge_sort_O3(x, 0, N-1);
  // print_list(x, N);
  
  return 0;
}
